esphome/components/rtttl/rtttl.cpp

```diff
@@ -0,0 +1,184 @@
+#include "rtttl.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace rtttl {
+
+static const char* TAG = "rtttl";
+
+// These values can also be found as constants in the Tone library (Tone.h)
+static const uint16_t NOTES[] = {0,    262,  277,  294,  311,  330,  349,  370,  392,  415,  440,  466,  494,
+                                 523,  554,  587,  622,  659,  698,  740,  784,  831,  880,  932,  988,  1047,
+                                 1109, 1175, 1245, 1319, 1397, 1480, 1568, 1661, 1760, 1865, 1976, 2093, 2217,
+                                 2349, 2489, 2637, 2794, 2960, 3136, 3322, 3520, 3729, 3951};
+
+inline static void skip_space_or_char(std::string::const_iterator& it, char c) {
+  while (*it == c || *it == ' ')
+    it++;
+}
+
+inline static uint8_t get_integer(std::string::const_iterator& it) {
+  uint8_t ret = 0;
+  while (isdigit(*it)) {
+    ret = (ret * 10) + (*it++ - '0');
+  }
+  return ret;
+}
+
+void Rtttl::dump_config() { ESP_LOGCONFIG(TAG, "Rtttl"); }
+
+void Rtttl::play(std::string rtttl) {
+  this->rtttl_ = std::move(rtttl);
+
+  ESP_LOGD(TAG, "Playing song %s", rtttl_.c_str());
+
+  p_ = rtttl_.cbegin();
+
+  this->default_duration_ = 4;
+  this->default_octave_ = 6;
+  int bpm = 63;
+  uint8_t num;
+
+  // format: d=N,o=N,b=NNN:
+  // find the start (skip name, etc)
+
+  // ignore name
+  while (*p_ != ':')
+    p_++;
+
+  skip_space_or_char(p_, ':');
+
+  // get default duration
+  if (*p_ == 'd') {
+    p_++;
+    p_++;  // skip "d="
+    num = get_integer(p_);
+    if (num > 0)
+      default_duration_ = num;
+    skip_space_or_char(p_, ',');
+  }
+
+  // get default octave
+  if (*p_ == 'o') {
+    p_++;
+    p_++;  // skip "o="
+    num = get_integer(p_);
+    if (num >= 3 && num <= 7)
+      default_octave_ = num;
+    skip_space_or_char(p_, ',');
+  }
+
+  // get BPM
+  if (*p_ == 'b') {
+    p_++;
+    p_++;  // skip "b="
+    num = get_integer(p_);
+    if (num != 0)
+      bpm = num;
+    skip_space_or_char(p_, ':');
+  }
+
+  // BPM usually expresses the number of quarter notes per minute
+  this->wholenote_ = (60 * 1000L / bpm) * 4;  // this is the time for whole note (in milliseconds)
```

Comment:
Can the multiplication by 4 be put before the division? Like this:

```suggestion
  this->wholenote_ = (60 * 1000L * 4) / bpm;  // this is the time for whole note (in milliseconds)
```

Then we have better rounding.